vc/vcjwt.go

```diff
@@ -0,0 +1,146 @@
+package vc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/tbd54566975/web5-go/jwt"
+)
+
+// Verify verifies the decoded vc-jwt. It checks for the presence of required fields and verifies the jwt.
+// It returns the decoded vc-jwt and an error if the verification fails.
+func Verify[T CredentialSubject](vcJWT string) (DecodedVCJWT[T], error) {
+	decoded, err := Decode[T](vcJWT)
+	if err != nil {
+		return decoded, err
+	}
+
+	return decoded, decoded.Verify()
+}
+
+// Decode decodes a vc-jwt as per the [spec] and returns [Decoded].
+// Note: this function uses certain fields from the jwt claims to eagrly populate the vc model as described
+// in the encoding section of the spec. The jwt fields will clobber any values that exist in the vc model.
+// While the jwt claims should match the counterpart values in the vc model, it's possible that they don't
+// but there would be no way to know if they don't match given that they're overwritten.
+//
+// [spec]: https://www.w3.org/TR/vc-data-model/#json-web-token
+func Decode[T CredentialSubject](vcJWT string) (DecodedVCJWT[T], error) {
+	decoded, err := jwt.Decode(vcJWT)
+	if err != nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("failed to decode vc-jwt: %w", err)
+	}
+
+	if decoded.Claims.Misc == nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("vc-jwt missing vc claim")
+	}
+
+	if _, ok := decoded.Claims.Misc["vc"]; ok == false {
+		return DecodedVCJWT[T]{}, fmt.Errorf("vc-jwt missing vc claim")
+	}
+
+	bytes, err := json.Marshal(decoded.Claims.Misc["vc"])
+	if err != nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("failed to decode vc claim: %w", err)
+	}
+
+	var vc DataModel[T]
+	if err := json.Unmarshal(bytes, &vc); err != nil {
+		return DecodedVCJWT[T]{}, fmt.Errorf("failed to decode vc claim: %w", err)
+	}
+
+	// the following conditionals are included to conform with the jwt decoding section
+	// of the specification defined here: https://www.w3.org/TR/vc-data-model/#jwt-decoding
+	if decoded.Claims.Issuer != "" {
+		vc.Issuer = decoded.Claims.Issuer
+	}
+
+	if decoded.Claims.JTI != "" {
+		vc.ID = decoded.Claims.JTI
+	}
+
+	if decoded.Claims.Subject != "" {
+		vc.CredentialSubject.SetID(decoded.Claims.Subject)
+	}
+
+	if decoded.Claims.Expiration != 0 {
+		vc.ExpirationDate = time.Unix(decoded.Claims.Expiration, 0).UTC().Format(time.RFC3339)
+	}
+
+	if decoded.Claims.NotBefore != 0 {
+		vc.IssuanceDate = time.Unix(decoded.Claims.NotBefore, 0).UTC().Format(time.RFC3339)
+	}
+
+	return DecodedVCJWT[T]{
+		JWT: decoded,
+		VC:  vc,
+	}, nil
+
+}
+
+// DecodedVCJWT represents a decoded vc-jwt. It contains the decoded jwt and decoded vc data model
+type DecodedVCJWT[T CredentialSubject] struct {
+	JWT jwt.Decoded
+	VC  DataModel[T]
+}
+
+// Verify verifies the decoded vc-jwt. It checks for the presence of required fields and verifies the jwt.
+func (vcjwt DecodedVCJWT[T]) Verify() error {
+	if vcjwt.VC.Issuer == "" {
```

Comment:
Thoughts about putting all conditions in this function to `vcjwt.VC` into `vc.go`? That is to say, create a function in `vc.go` like...

```go
func (vc DataModel[T]) Verify() error {
	// these things
}
```